internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -395,40 +382,64 @@ func calculateStatus(allMSs []*clusterv1.MachineSet, newMS *clusterv1.MachineSet
 	}
 
 	// Calculate the label selector. We check the error in the MD reconcile function, ignore here.
-	selector, _ := metav1.LabelSelectorAsSelector(&deployment.Spec.Selector)
-
-	status := clusterv1.MachineDeploymentStatus{
-		// TODO: Ensure that if we start retrying status updates, we won't pick up a new Generation value.
-		ObservedGeneration:  deployment.Generation,
-		Selector:            selector.String(),
-		Replicas:            mdutil.GetActualReplicaCountForMachineSets(allMSs),
-		UpdatedReplicas:     mdutil.GetActualReplicaCountForMachineSets([]*clusterv1.MachineSet{newMS}),
-		ReadyReplicas:       mdutil.GetReadyReplicaCountForMachineSets(allMSs),
-		AvailableReplicas:   availableReplicas,
-		UnavailableReplicas: unavailableReplicas,
-		Conditions:          deployment.Status.Conditions,
-	}
+	selector, _ := metav1.LabelSelectorAsSelector(&machineDeployment.Spec.Selector)
 
-	if *deployment.Spec.Replicas == status.ReadyReplicas {
-		status.Phase = string(clusterv1.MachineDeploymentPhaseRunning)
+	if desiredReplicas == readyReplicas {
+		phase = string(clusterv1.MachineDeploymentPhaseRunning)
+		conditions.MarkTrue(machineDeployment, clusterv1.ResizedCondition)
 	}
-	if *deployment.Spec.Replicas > status.ReadyReplicas {
-		status.Phase = string(clusterv1.MachineDeploymentPhaseScalingUp)
+	if desiredReplicas > readyReplicas {
+		phase = string(clusterv1.MachineDeploymentPhaseScalingUp)
+		conditions.MarkFalse(machineDeployment, clusterv1.ResizedCondition, clusterv1.ScalingUpReason, clusterv1.ConditionSeverityWarning,
+			"Scaling up MachineDeployment to %d replicas (actual %d)", desiredReplicas, readyReplicas)
 	}
 	// This is the same as unavailableReplicas, but we have to recalculate because unavailableReplicas
 	// would have been reset to zero above if it was negative
 	if totalReplicas-availableReplicas < 0 {
-		status.Phase = string(clusterv1.MachineDeploymentPhaseScalingDown)
+		phase = string(clusterv1.MachineDeploymentPhaseScalingDown)
+		conditions.MarkFalse(machineDeployment, clusterv1.ResizedCondition, clusterv1.ScalingDownReason, clusterv1.ConditionSeverityWarning,
+			"Scaling down MachineDeployment to %d replicas (actual %d)", desiredReplicas, availableReplicas)
 	}
+
+	msConditionsGetterList := make([]conditions.Getter, 0, len(allMSs))
 	for _, ms := range allMSs {
 		if ms != nil {
 			if ms.Status.FailureReason != nil || ms.Status.FailureMessage != nil {
-				status.Phase = string(clusterv1.MachineDeploymentPhaseFailed)
+				phase = string(clusterv1.MachineDeploymentPhaseFailed)
 				break
 			}
 		}
+		msConditionsGetterList = append(msConditionsGetterList, ms)
```

Comment:
Given that there is a nil check in l.406 it seems to be me that there is a real possibility of having a `nil` in allMSs. If I see correctly adding a nil to `msConditionsGetterList` will lead to a panic somewhere in `SetAggregate` (I think when `GetConditions` is eventually called).

Should we use something like:
```go
		if ms == nil {
                    continue
                 }
```
to safely filter out the nils?

Reply:
Good point, I'd expect empty to be a valid case and SetAggregate to make the right decision on what to set in the aggregated Ready or just not set it. See if others can confirm.

Reply:
Good point, might be that we want to use the nil for the calculation.

So the question is what should `SetAggregate` do when one of the ConditionGetters is nil.

Reply:
@enxebre Another question. When we modify MachineDeployments we keep old MachineSets (RevisionHistoryLimit default: 1). It looks like that MachineSets with 0 replicas are not ready: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1500823799287779328/artifacts/clusters/bootstrap/resources/k8s-upgrade-and-conformance-9lsgih/MachineSet/k8s-upgrade-and-conformance-oskt1y-md-0-qnd9f-69f47b7dbc.yaml

I"m not sure how the aggregation works, but wouldn't this mean that as soon as we have an old MachineSet with 0 replicas (which we should almost always have after the second MD rollout) MachineSetsReady will always be false?

Reply:
Again good point @sbueringer. I'd expect the correct behaviour when MachineSet.replicas=0 to be the following:

1:
`conditions.SetAggregate(ms, clusterv1.MachinesReadyCondition, collections.FromMachines(filteredMachines...).ConditionGetters(), conditions.AddSourceRef(), conditions.WithStepCounterIf(false))` -> No target condition is generated since there's no machines

2:
`
	conditions.SetSummary(machineSet,
		conditions.WithConditions(
			clusterv1.MachinesCreatedCondition,
			clusterv1.ResizedCondition,
			clusterv1.MachinesReadyCondition,
		),
	)
`

Sets ready=true by ignoring MachinesCreatedCondition/MachinesReadyCondition and honouring ResizedCondition.
Then this gets bubbled up to MachineDeployment.

The works as expected atm when a MachineSet is born with replicas=0. However as you pointed out when a MachineSet goes from replicas>1 to 0 it ends up with a permanent Machine deleted condition
```
  conditions:
  - lastTransitionTime: "2022-03-09T11:23:57Z"
    message: 1 of 2 completed
    reason: Deleted @ Machine/agl-w7pf6-agl-agl-custom-674bbf874c-6pxxj
    severity: Info
    status: "False"
    type: Ready
```

I guess this is because the aws provider sets the infrastructureReady condition as false on deletion https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controllers/awsmachine_controller.go#L350, So then this is bubbled up to the MachineSet and never cleaned up as eventually the Machine CR goes away and the MachineSet controller goes no-op on reconcile.

I consider this bug caused by a misuse of the infraReady condition by the AWS provider which should not mark infraReady as false during a delete operation.

cc @fabriziopandini thoughts?

Reply:
Possible path forward (untested) https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3303. This bugfix would be though technically an API breaking change so we might need to think on how to transition if we agree on this being the path forward